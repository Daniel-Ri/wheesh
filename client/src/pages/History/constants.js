export const GET_ORDER = 'History/GET_ORDER';
export const SET_ORDER = 'History/SET_ORDER';
