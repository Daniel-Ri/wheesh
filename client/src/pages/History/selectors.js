import { createSelector } from 'reselect';
import { initialState } from './reducer';

const selectHistoryState = (state) => state.history || initialState;

export const selectOrder = createSelector(selectHistoryState, (state) => state.order);
