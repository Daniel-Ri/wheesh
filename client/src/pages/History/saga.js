import { getOrder } from '@domain/api';
import { takeLatest, call, put } from 'redux-saga/effects';
import { setOrder } from './actions';
import { GET_ORDER } from './constants';

function* doGetOrder({ orderId, handleError }) {
  try {
    const response = yield call(getOrder, orderId);
    yield put(setOrder(response.data));
  } catch (error) {
    handleError(error.response.data.message);
  }
}

export default function* historySaga() {
  yield takeLatest(GET_ORDER, doGetOrder);
}
