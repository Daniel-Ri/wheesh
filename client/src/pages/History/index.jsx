import { connect, useDispatch } from 'react-redux';
import PropTypes from 'prop-types';
import { FormattedMessage, injectIntl } from 'react-intl';
import { useNavigate, useParams } from 'react-router-dom';

import arrowImage from '@static/images/arrowTrain.png';
import toast from 'react-hot-toast';
import { useEffect } from 'react';
import BackBtn from '@components/BackBtn';
import { createStructuredSelector } from 'reselect';
import { formatDateWithDay, formatHour, formatOrderDate, formatRupiah } from '@utils/handleValue';
import { selectLocale } from '@containers/App/selectors';
import { getOrder, setOrder } from './actions';
import { selectOrder } from './selectors';
import OrderedSeatCard from './components/OrderedSeatCard';

import classes from './style.module.scss';

const History = ({ order, locale, intl: { formatMessage } }) => {
  const { orderId } = useParams();

  const dispatch = useDispatch();
  const navigate = useNavigate();

  const handleErrorGetOrder = (errorMsg) => {
    toast.error(errorMsg);
    navigate('/my-tickets');
  };

  useEffect(() => {
    dispatch(getOrder(orderId, handleErrorGetOrder));
    // eslint-disable-next-line react-hooks/exhaustive-deps
  }, [orderId]);

  useEffect(() => {
    if (!order) return;

    if (!order.Payment.isPaid) {
      toast.error(formatMessage({ id: 'app_havent_paid_this_order' }));
      navigate(`/unpaid/${orderId}`);
      return;
    }

    if (new Date(order.Schedule.arrivalTime) >= new Date(new Date().getTime() - 6 * 60 * 60 * 1000)) {
      navigate(`/order/${orderId}`);
      return;
    }

    return () => {
      dispatch(setOrder(null));
    };
    // eslint-disable-next-line react-hooks/exhaustive-deps
  }, [order]);

  return (
    <main data-testid="History" className={classes.main}>
      <div className={classes.container}>
        <header>
          <BackBtn handleClickBack={() => navigate('/my-tickets')} />
          <h1>
            <FormattedMessage id="app_history" />
          </h1>
        </header>

        <section className={classes.metaData}>
          <div>
            <FormattedMessage id="app_order_number" />: {order?.id}
          </div>
          <div>
            <FormattedMessage id="app_order_time" />: {formatHour(order?.createdAt)} {formatOrderDate(order?.createdAt)}
          </div>
        </section>

        <section>
          <div className={classes.sectionDesc}>
            <div className={classes.row}>
              <div className={classes.dateTime}>{formatDateWithDay(order?.Schedule.departureTime, locale)}</div>
            </div>
            <div className={classes.row}>
              <div className={classes.timeAndPlace}>
                <div className={classes.time}>
                  {formatHour(order?.Schedule.departureTime)}
                  <sup> WIB</sup>
                </div>
                <div className={classes.place}>{order?.Schedule.departureStation.name}</div>
              </div>
              <div className={classes.train}>
                <div className={classes.name}>{order?.Schedule.Train.name}</div>
                <div className={classes.arrowImage}>
                  <img src="../../static/images/arrowTrain.png" alt="" />
                </div>
              </div>
              <div className={classes.timeAndPlace}>
                <div className={classes.time}>
                  {formatHour(order?.Schedule.arrivalTime)}
                  <sup> WIB</sup>
                </div>
                <div className={classes.place}>{order?.Schedule.arrivalStation.name}</div>
              </div>
            </div>
          </div>
        </section>

        <section>
          <div className={classes.header}>
            <FormattedMessage id="app_passenger" />
          </div>
          <hr />
          <div className={classes.sectionDesc}>
            {order?.OrderedSeats.map((orderedSeat, idx) => (
              <OrderedSeatCard
                key={orderedSeat.id}
                orderedSeat={orderedSeat}
                addLine={idx !== order.OrderedSeats.length - 1}
              />
            ))}
          </div>
        </section>

        <section>
          <div className={classes.sectionDesc}>
            <div className={classes.totalPrice}>
              <div>
                <FormattedMessage id="app_total_payment_amount" />
              </div>
              <div>{formatRupiah(order?.Payment.amount)}</div>
            </div>
          </div>
        </section>
      </div>
    </main>
  );
};

History.propTypes = {
  order: PropTypes.object,
  locale: PropTypes.string.isRequired,
  intl: PropTypes.object,
};

const mapStateToProps = createStructuredSelector({
  order: selectOrder,
  locale: selectLocale,
});

export default injectIntl(connect(mapStateToProps)(History));
