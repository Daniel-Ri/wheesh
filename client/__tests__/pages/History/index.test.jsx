import { render } from '@utils/testHelper';
import History from '@pages/History';

const mockNavigate = jest.fn();

// jest.mock('react-redux', () => ({
//   useDispatch: jest.fn(),
// }));

jest.mock('react-router-dom', () => ({
  useNavigate: () => mockNavigate,
  useParams: () => ({
    orderId: 9,
  }),
}));

let wrapper;

const mockProps = {
  order: {
    id: 9,
    userId: 2,
    scheduleId: 1,
    isNotified: true,
    createdAt: '2023-12-20T06:02:10.000Z',
    Payment: {
      id: 9,
      orderId: 9,
      amount: 600000,
      isPaid: true,
      duePayment: '2023-12-20T06:02:11.000Z',
    },
    Schedule: {
      id: 1,
      trainId: 1,
      departureStationId: 1,
      arrivalStationId: 3,
      departureTime: new Date(new Date().getTime() - 8 * 60 * 60 * 1000),
      arrivalTime: new Date(new Date().getTime() - 7 * 60 * 60 * 1000),
      departureStation: {
        id: 1,
        name: 'Halim',
      },
      arrivalStation: {
        id: 3,
        name: 'Padalarang',
      },
      Train: {
        id: 1,
        name: 'G1201',
      },
    },
    OrderedSeats: [
      {
        id: 1721,
        orderId: 9,
        seatId: 361,
        price: 200000,
        gender: 'Male',
        dateOfBirth: '1951-08-23T01:28:44.000Z',
        idCard: '1217434274792992',
        name: 'John Ganteng',
        email: '1buyf4bn@example.com',
        Seat: {
          id: 361,
          carriageId: 5,
          seatNumber: '11D',
          seatClass: 'economy',
          Carriage: {
            id: 5,
            trainId: 1,
            carriageNumber: 5,
          },
        },
      },
      {
        id: 1722,
        orderId: 9,
        seatId: 169,
        price: 200000,
        gender: 'Male',
        dateOfBirth: '1996-05-31T07:27:29.000Z',
        idCard: '1332443674346541',
        name: 'Bob Johnson',
        email: null,
        Seat: {
          id: 169,
          carriageId: 3,
          seatNumber: '9B',
          seatClass: 'economy',
          Carriage: {
            id: 3,
            trainId: 1,
            carriageNumber: 3,
          },
        },
      },
      {
        id: 1723,
        orderId: 9,
        seatId: 451,
        price: 200000,
        gender: 'Male',
        dateOfBirth: '1985-06-15T03:40:42.000Z',
        idCard: '4930380435545503',
        name: 'Alice Smith',
        email: 'thvw0o17@example.com',
        Seat: {
          id: 451,
          carriageId: 6,
          seatNumber: '11D',
          seatClass: 'economy',
          Carriage: {
            id: 6,
            trainId: 1,
            carriageNumber: 6,
          },
        },
      },
    ],
  },
  locale: 'id',
};

beforeEach(() => {
  wrapper = render(<History {...mockProps} />);
});

describe('Page History', () => {
  // test('Renders History page', () => {
  //   const { getByTestId } = wrapper;
  //   expect(getByTestId('History')).toBeInTheDocument();
  // });

  // test('Should match snapshot', () => {
  //   expect(wrapper).toMatchSnapshot();
  // });
  test('test', () => {});
});
