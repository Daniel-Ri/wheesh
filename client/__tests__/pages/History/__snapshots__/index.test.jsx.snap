// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Page History Should match snapshot 1`] = `
{
  "asFragment": [Function],
  "baseElement": <body>
    <div>
      <main
        class="main"
        data-testid="History"
      >
        <div
          class="container"
        >
          <header>
            <button
              class="MuiButtonBase-root MuiButton-root MuiButton-text MuiButton-textPrimary MuiButton-sizeMedium MuiButton-textSizeMedium MuiButton-root MuiButton-text MuiButton-textPrimary MuiButton-sizeMedium MuiButton-textSizeMedium btn css-1e6y48t-MuiButtonBase-root-MuiButton-root"
              data-testid="BackBtn"
              tabindex="0"
              type="button"
            >
              Kembali
              <span
                class="MuiTouchRipple-root css-8je8zh-MuiTouchRipple-root"
              />
            </button>
            <h1>
              Riwayat
            </h1>
          </header>
          <section
            class="metaData"
          >
            <div>
              Nomor Pesanan
              : 
            </div>
            <div>
              Waktu Pesanan
              : 
              Invalid Date
               
              NaN/NaN/NaN
            </div>
          </section>
          <section>
            <div
              class="sectionDesc"
            >
              <div
                class="row"
              >
                <div
                  class="dateTime"
                >
                  undefined, NaN Invalid Date NaN
                </div>
              </div>
              <div
                class="row"
              >
                <div
                  class="timeAndPlace"
                >
                  <div
                    class="time"
                  >
                    Invalid Date
                    <sup>
                       WIB
                    </sup>
                  </div>
                  <div
                    class="place"
                  />
                </div>
                <div
                  class="train"
                >
                  <div
                    class="name"
                  />
                  <div
                    class="arrowImage"
                  >
                    <img
                      alt=""
                      src="../../static/images/arrowTrain.png"
                    />
                  </div>
                </div>
                <div
                  class="timeAndPlace"
                >
                  <div
                    class="time"
                  >
                    Invalid Date
                    <sup>
                       WIB
                    </sup>
                  </div>
                  <div
                    class="place"
                  />
                </div>
              </div>
            </div>
          </section>
          <section>
            <div
              class="header"
            >
              Penumpang
            </div>
            <hr />
            <div
              class="sectionDesc"
            />
          </section>
          <section>
            <div
              class="sectionDesc"
            >
              <div
                class="totalPrice"
              >
                <div>
                  Jumlah Total Pembayaran
                </div>
                <div>
                  RpNaN
                </div>
              </div>
            </div>
          </section>
        </div>
      </main>
    </div>
  </body>,
  "container": <div>
    <main
      class="main"
      data-testid="History"
    >
      <div
        class="container"
      >
        <header>
          <button
            class="MuiButtonBase-root MuiButton-root MuiButton-text MuiButton-textPrimary MuiButton-sizeMedium MuiButton-textSizeMedium MuiButton-root MuiButton-text MuiButton-textPrimary MuiButton-sizeMedium MuiButton-textSizeMedium btn css-1e6y48t-MuiButtonBase-root-MuiButton-root"
            data-testid="BackBtn"
            tabindex="0"
            type="button"
          >
            Kembali
            <span
              class="MuiTouchRipple-root css-8je8zh-MuiTouchRipple-root"
            />
          </button>
          <h1>
            Riwayat
          </h1>
        </header>
        <section
          class="metaData"
        >
          <div>
            Nomor Pesanan
            : 
          </div>
          <div>
            Waktu Pesanan
            : 
            Invalid Date
             
            NaN/NaN/NaN
          </div>
        </section>
        <section>
          <div
            class="sectionDesc"
          >
            <div
              class="row"
            >
              <div
                class="dateTime"
              >
                undefined, NaN Invalid Date NaN
              </div>
            </div>
            <div
              class="row"
            >
              <div
                class="timeAndPlace"
              >
                <div
                  class="time"
                >
                  Invalid Date
                  <sup>
                     WIB
                  </sup>
                </div>
                <div
                  class="place"
                />
              </div>
              <div
                class="train"
              >
                <div
                  class="name"
                />
                <div
                  class="arrowImage"
                >
                  <img
                    alt=""
                    src="../../static/images/arrowTrain.png"
                  />
                </div>
              </div>
              <div
                class="timeAndPlace"
              >
                <div
                  class="time"
                >
                  Invalid Date
                  <sup>
                     WIB
                  </sup>
                </div>
                <div
                  class="place"
                />
              </div>
            </div>
          </div>
        </section>
        <section>
          <div
            class="header"
          >
            Penumpang
          </div>
          <hr />
          <div
            class="sectionDesc"
          />
        </section>
        <section>
          <div
            class="sectionDesc"
          >
            <div
              class="totalPrice"
            >
              <div>
                Jumlah Total Pembayaran
              </div>
              <div>
                RpNaN
              </div>
            </div>
          </div>
        </section>
      </div>
    </main>
  </div>,
  "debug": [Function],
  "findAllByAltText": [Function],
  "findAllByDisplayValue": [Function],
  "findAllByLabelText": [Function],
  "findAllByPlaceholderText": [Function],
  "findAllByRole": [Function],
  "findAllByTestId": [Function],
  "findAllByText": [Function],
  "findAllByTitle": [Function],
  "findByAltText": [Function],
  "findByDisplayValue": [Function],
  "findByLabelText": [Function],
  "findByPlaceholderText": [Function],
  "findByRole": [Function],
  "findByTestId": [Function],
  "findByText": [Function],
  "findByTitle": [Function],
  "getAllByAltText": [Function],
  "getAllByDisplayValue": [Function],
  "getAllByLabelText": [Function],
  "getAllByPlaceholderText": [Function],
  "getAllByRole": [Function],
  "getAllByTestId": [Function],
  "getAllByText": [Function],
  "getAllByTitle": [Function],
  "getByAltText": [Function],
  "getByDisplayValue": [Function],
  "getByLabelText": [Function],
  "getByPlaceholderText": [Function],
  "getByRole": [Function],
  "getByTestId": [Function],
  "getByText": [Function],
  "getByTitle": [Function],
  "queryAllByAltText": [Function],
  "queryAllByDisplayValue": [Function],
  "queryAllByLabelText": [Function],
  "queryAllByPlaceholderText": [Function],
  "queryAllByRole": [Function],
  "queryAllByTestId": [Function],
  "queryAllByText": [Function],
  "queryAllByTitle": [Function],
  "queryByAltText": [Function],
  "queryByDisplayValue": [Function],
  "queryByLabelText": [Function],
  "queryByPlaceholderText": [Function],
  "queryByRole": [Function],
  "queryByTestId": [Function],
  "queryByText": [Function],
  "queryByTitle": [Function],
  "rerender": [Function],
  "unmount": [Function],
}
`;
